### DISCUSSION TITLE
Search component

### DISCUSSION THREAD

**Discussion by sang-yeo** [twilio-labs/paste#677] [Q&A] [answered]
We are planning to add a Widget search feature in Studio. It's similar to Chrome search feature. Could you recommend which Paste component I need to use for this feature? 

Basic interaction will be

1. A user clicks the search icon from Canvas Controller 
2. Search input(?) pops up (Which component should I use? I planning to use an input component without label) 
Thanks!

**Comment by serifluous** [comment, marked as answer]
Thanks, @sang-yeo ! Could you share a screenshot, paper sketch, or link of what you're trying to achieve?

Off the top of my head, a constraint to consider is that you won't be able to turn the Button component (the search icon) into an Input on click. But what you can do is have a very short Input that expands on focus (like the search in console nav) OR open a modal or other dialog that has an input.

[Input](https://paste.twilio.design/form-elements/input/) is likely the right component to use here. But if you need typeahead/autocomplete (like the search in console nav or on the Paste doc site), you'll want to use [Combobox](https://paste.twilio.design/components/combobox)—we don't have a Sketch symbol for it yet, but you can use the Select symbol for now and pair it with the Menu symbol.

Search is generally the only place where we'd say not including a label is ok, but it really depends on a few things:
1. Can the UI actually allow for it? If you can place a label, it helps make it that much clearer to all users.
2. If you decide that it doesn't, take a look at the [accessibility information on form labels](https://paste.twilio.design/form-elements/input/#accessibility) and work with your engineering team to make sure a non-visible form label can be read by screen readers. Another way to do this is include a [search icon prefix](https://paste.twilio.design/form-elements/input/#input-with-add-ons-prefix-suffix-text-or-icons) on the form field and [give the icon a descriptive title](https://paste.twilio.design/icon-system#specify-if-an-icon-is-decorative) that tells a user what the input is for.

**Reply by sang-yeo** [reply]
Sure, here is what I'm thinking [click here](https://drive.google.com/file/d/1O5rg2LDdDSQATYU30enGKzglDcba-K2L/view?usp=sharing)

**Reply by sang-yeo** [reply]
@serifluous

**Reply by sang-yeo** [reply]
2nd concept looks like this [click here](https://drive.google.com/file/d/1u13RXZqhERw0IkkTb_VMC9bUw26k1FzJ/view?usp=sharing)

**Reply by serifluous** [reply]
The first concept won't be doable since it runs into the constraint I mentioned about not being able to turn a Button into an Input.

The second concept should be doable. There's definitely room for a visible label in the popover, too.

**Reply by sang-yeo** [reply]
Thanks for answering! is your suggestion adding a visible label and removing the placeholder?

**Reply by serifluous** [reply]
Yep, that would work for Concept 2.

We discussed a little more in the team, and if you were interested in getting Concept 1 to work, what you could do is have the search icon button always visible, and it would toggle whether the search input is hidden or visible. So instead of Button --> Input, instead you'd do Button --> Button + Input.

**Comment by SiTaggart** [comment]
One thing that will affect which input control you use will be the search behavior itself. Can you describe what happens when I type in the search input @sang-yeo?

**Reply by sang-yeo** [reply]
We are thinking [this](https://drive.google.com/file/d/1jOYiOH19bq-fOQQi08oHa3ybcARGBSE8/view?usp=sharing), basically display the name and type of Widget that a customer searches. @SiTaggart

**Reply by sang-yeo** [reply]
Also, which state do I need to use for no result?

**Reply by SiTaggart** [reply]
So the component you need is a Combobox.

For no results you have 2 choices:

1. Don't show any option
2. Show one option that states there are no returned results.

The design of number 2 is up to you right now. We don't have any opinions on that at this moment.

**Reply by sang-yeo** [reply]
This is great! thank you so much. One more question. Can I add "X" icon inside of input so customer can delete the input? Like [this](https://drive.google.com/file/d/1RRENxGTMbHgxwMLzPxtmRaliJXQqzGIF/view?usp=sharing)

**Reply by SiTaggart** [reply]
Not exactly like that, no. You can use the input suffix to add a clear icon though https://paste.twilio.design/components/combobox#combobox-with-add-ons-prefix-suffix-text-or-icons

**Reply by sang-yeo** [reply]
Great. I applied your suggestion to the mock[here](https://drive.google.com/file/d/1-dLlk7mcHYnW3ZFFNYyuLaOBSePCtHdD/view?usp=sharing). Thanks again for detailed feedback

**Reply by serifluous** [reply]
Looks good 👍 .

If you still end up doing some exploration on the clear icon, you can also take a look at the code sandbox link on the answer to discussion #605 for some options.

**Reply by aayushpi** [reply]
@sang-yeo Recommend setting the initial focus on the popover to the combobox instead of the close, since that is what the user is optimistically trying to reach.

**Reply by sang-yeo** [reply]
@aayushpi will do. Thank you! This is really helpful.
